libraries/libduck/DirectoryEntry.cpp

```diff
@@ -19,7 +19,149 @@
 
+#include <sys/stat.h>
+#include <sstream>
+#include <iomanip>
+
 using namespace Duck;
 
-DirectoryEntry::DirectoryEntry(const Path& parent_path, const struct dirent* entry):
-	m_path(parent_path / entry->d_name), m_inode(entry->d_ino), m_type((Type) entry->d_type), m_name(entry->d_name) {}
\ No newline at end of file
+static constexpr auto REG_FORMAT = "\033[39m";
+static constexpr auto DIR_FORMAT = "\033[36m";
+static constexpr auto BLK_FORMAT = "\033[33m";
+static constexpr auto CHR_FORMAT = "\033[32m";
+static constexpr auto LNK_FORMAT = "\033[34m";
+static constexpr auto FIFO_FORMAT = "\033[35m";
+static constexpr auto SOCK_FORMAT = "\033[38m";
+
+static constexpr auto R_FORMAT = "\033[36m";
+static constexpr auto W_FORMAT = "\033[31m";
+static constexpr auto X_FORMAT = "\033[32m";
+
+static constexpr auto RESET_FORMAT = REG_FORMAT; // TODO "\033[0m";
+
+static constexpr char entry_type_char(Duck::DirectoryEntry::Type d_type) {
+	char type = '-';
+	if(d_type == Duck::DirectoryEntry::DIRECTORY)
+		type = 'd';
+	else if(d_type == Duck::DirectoryEntry::BLOCKDEVICE)
+		type = 'b';
+	else if(d_type == Duck::DirectoryEntry::CHARDEVICE)
+		type = 'c';
+	else if(d_type == Duck::DirectoryEntry::SYMLINK)
+		type = 'l';
+
+	return type;
+}
+
+static std::string entry_permission_string(mode_t mode, mode_t bit, char const* color, char c) {
+	std::string s;
+	if((mode & bit) != 0) {
+		if(color) {
+			s += color;
+		}
+		s += c;
+		if(color) {
+			s += RESET_FORMAT;
+		}
+	} else {
+		if(color) {
+			s += RESET_FORMAT;
+		}
+		s += '-';
+	}
+	return s;
+}
+
+static std::string entry_permissions_string(mode_t mode, bool colorize) {
+	std::string s;
+	s += entry_permission_string(mode, S_IRUSR, colorize ? R_FORMAT : nullptr, 'r');
+	s += entry_permission_string(mode, S_IWUSR, colorize ? W_FORMAT : nullptr, 'w');
+	s += entry_permission_string(mode, S_IXUSR, colorize ? X_FORMAT : nullptr, 'x');
+	s += entry_permission_string(mode, S_IRGRP, colorize ? R_FORMAT : nullptr, 'r');
+	s += entry_permission_string(mode, S_IWGRP, colorize ? W_FORMAT : nullptr, 'w');
+	s += entry_permission_string(mode, S_IXGRP, colorize ? X_FORMAT : nullptr, 'x');
+	s += entry_permission_string(mode, S_IROTH, colorize ? R_FORMAT : nullptr, 'r');
+	s += entry_permission_string(mode, S_IWOTH, colorize ? W_FORMAT : nullptr, 'w');
+	s += entry_permission_string(mode, S_IXOTH, colorize ? X_FORMAT : nullptr, 'x');
+	return s;
+}
+
+static std::string entry_fmt_size(Duck::DirectoryEntry::Type d_type, off_t size) {
+	std::stringstream ss;
+	if(d_type == Duck::DirectoryEntry::REGULAR) {
+		char mul = ' ';
+		if(size >= 1024 * 1024 * 1024) {
+			mul = 'G';
+			size /= 1024 * 1024 * 1024;
+		} else if(size >= 1024 * 1024) {
+			mul = 'M';
+			size /= 1024 * 1024;
+		} else if(size >= 1024) {
+			mul = 'K';
+			size /= 1024;
+		}
+		ss << std::right << std::setw(4) << size << mul;
+	} else {
+		ss << std::right << std::setw(4) << '-' << ' ';
+	}
+	return ss.str();
+}
```

Comment:
There's actually logic for this sort of thing in `Memory::Amount` in libsys, although that is technically for memory and not storage sizes (not that it makes a ton of difference). We can leave this for now, but I think later I'll make a general utility class for formatting sizes like this